charger/easee.go

```diff
@@ -481,6 +482,27 @@ func (c *Easee) waitForTickResponse(expectedTick int64) error {
 	}
 }
 
+// wait for up to 3s for current become 32A
+func (c *Easee) waitForDynamicChargerMaxA() {
+	timer := time.NewTimer(3 * time.Second)
+	for {
```

Comment:
Wäre es nicht besser, aus dem `respChan` zu lesen anstatt eines busy waits? Setzt natürlich voraus, dass sonst niemand draus liest, aber das sollte hier gegeben sein.

Reply:
Eine Lösung mit Channel hatte ich zwischenzeitlich auch gebaut. Siehe https://github.com/evcc-io/evcc/pull/8588/commits/c1d468eebd0f3cc3ccd53add6c2be5ba899415f6

Das Problem ist aber, daß theoretisch das Product Update über die Go routine reinkommen könnte, während im Hauptthread der select noch nicht erreicht ist. Wir hätten das Update dann also verpasst, und wir würden bis zum Timeout warten. 
Das verpasste Update wiederum wollte ich mit einem Buffered Channel lösen. Der hilft aber auch nicht, weil wir ja bei "normalen" Dynamic Current updates dann auch in den Channel schreiben aber nicht davon lesen. Somit ist der Buffer immer voll. 
Nächste Idee war statt dem Buffer, einfach das warten auf Timeout zu verkürzen in dem zusätzlich alle 300ms geprüft wird ob sich `c.dynamicChargerCurrent` bereits verändert hat. Das löst das Problem mit dem verpassten Update, macht aber gleichzeitig den ganzen Channel Ansatz überflüssig. Vielleicht würden wir über der Channel noch mehr "in Echtzeit" von der Veränderung erfahren, aber für im Mittel 150ms die Komplexität zu erhöhen schien mir nicht sinnvoll. Keep it simple. :)

Ich war gedanklich schon soweit: Das Problem ließe sich mit einem blocking channel lösen. Dieser müßte dann aber aus enable() heraus dynamisch gesetzt und entfernt werden, so daß das Product Update immer nur dann published wenn auch jemand lauscht. Aber das wurde alles immer komplizierter und würde vermutlich mehr Kopfzerbrechen bereiten als es nützt.

Andere Frage: Du schreibst "busy wait" - der Timer ist doch aber nebenläufig, oder nicht? Der Sleep mit den 300ms am Ende ist nicht so ganz sexy. Ich könnte das noch auf einen Ticker umbauen, mir ist aber nicht klar ob das etwas bringt.

Reply:
> Das Problem ist aber, daß theoretisch das Product Update über die Go routine reinkommen könnte, während im Hauptthread der select noch nicht erreicht ist

dann lies doch einmal den Wert direkt aus?

Reply:
habe ich gemacht. Auch mit Lock. Aber es gibt natürlich immer noch diesen kleinen Moment zwischen unlock() und select in dem theoretisch der andere Thread das Product Update verarbeiten könnte. Ich gebe zu es ist unwahrscheinlich - aber möglich. Hängt halt vom scheduling ab. Ich habe keinen einfachen Weg gesehen das ganze 100%ig zuverlässig hinzubekommen. 

Was wir hier bräuchten ist ein Pattern das eine Kommunikation zwischen Hauptthread und Product Update Thread erlaubt - so in der Art: Hey, Product Update, warte mal kurz, ich müßte erst meinen Select starten. Ein blocking Channel würde das erfüllen. Aber jetzt kommt hinzu, daß wir das alles nur im Fall von Resume benötigen. Der Thread der die Product Updates verarbeitet kann nicht eigenständig erkennen ob wir uns gerade in so einer Situation befinden. 
Daher die oben erwähnte Idee, das man einen Blocking Channel wenn benötigt vom Hauptthread zum Product Update hoch reicht. Wenn ein Channel vorhanden ist, wird dieser dann blocking beschrieben. Sobald wir den Wert haben den wir brauchen machen signalisiert der Hauptthread das und wir schließen den Channel.
So etwas könnte man bauen. Es ist elegant, keine Frage - aber auch deutlich komplizierter als die aktuelle Lösung über den Poll alle paar hundert Millisekunden.

Reply:
Machs doch nicht so kompliziert und lass uns das erstmal ausprobieren. Fehlt noch im Code?

Reply:
Sag ich doch. :)
An dieser Stelle fehlt mAn nichts mehr. Ich resolve hier erstmal die Convo. Für den letzten Part oben schaue ich jetzt gleich nochmal was ich bauen kann. Möchte das wenn möglich heute Abend fertig bekommen.

Reply:
Den hätt ich gerne noch umgebaut. Wenn du keine Idee hast kann ich einen Vorschlag machen, dauert aber 2..3 Tage bis ich dazu komme. Ziel: busy wait weg, channel nutzen.

Reply:
> Sag ich doch. :)

Ich hatte verstanden du willst es ganz kompliziert machen. Mir würde existierenden Channel nutzen reichen.

Reply:
Da würde ich mich über einen Vorschlag von Dir freuen.
Den bestehenden Channel zu nutzen ist naheliegend. Ich habe leider nicht genug GoLang Erfahrung, um das Problem der Typisierung dann in den Griff zu bekommen. Für mich ist die Hürde, das über den aktuellen Channel  CommandResponses fließen, den Typ haben wir im ProductUpdate handler nicht.
Würdest Du diesen Teil des Problem mit [Type Switch](https://go.dev/doc/effective_go#type_switch) lösen?

Reply:
Vmtl ja -

Reply:
Habe heute nochmal drüber nachgedacht. Ich habe mal in einem subbranch hier eine Idee skizziert wie es mit einem flag und einem Channel funktionieren könnte. Vielleicht sind wir damit schon nah dran?

https://github.com/GrimmiMeloni/evcc/commit/76950c289d38c267edf893b31b34b690cd46101f